### ISSUE TITLE
Allow Kotlin Collections.List for @All qualifier

### ISSUE THREAD

**Issue by Dudeplayz** [quarkusio/quarkus#24766]
### Describe the bug

I was trying to inject different implementations of the same interface at the same time. Following the docs I have done it with:

```kotlin
@Inject
@All
protected lateinit var processors: List<IProcessor>
```

But on compilation it throws: 
```log
Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: javax.enterprise.inject.spi.DefinitionException: Wildcard is not a legal type argument for ...Resource#processors
```

As I am using Kotlin, I changed the type of the list to MutableList, which works.
I think `List` is the more appropriate type following the docs: `The injected instance is an immutable list of the contextual references of the disambiguated beans.`. So either both or only `List` should be supported. The later one could break existing programs.

I am also unsure, if the wildcard exception is the correct exception in this case?

### Expected behavior

An immutable list is injected for `List<T>`.

### Actual behavior

An exception is thrown.

### Quarkus version or git rev

2.8.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvn

**Comment by Dudeplayz** [comment]
Ok, but what is different for the MutableList? Is this a bug in Kotlin or wanted?

**Comment by Dudeplayz** [comment]
I have found this older ticket, stating to keep it as it is: https://youtrack.jetbrains.com/issue/KT-5792

**Comment by mkouba** [comment]
I don't think it's a bug in Kotlin. The point is that Kotlin's `List` is not `java.util.List` (it's read-only and does not translate 1:1 to `java.util.List`) and therefore you can't expect the `@All` qualifier to _just work_. Kotlin's `MutableList` on the other hand seems to behave like an alias for `java.util.List`.

@Ladicek and @evanchooly might have a better explanation ;-).

**Comment by Ladicek** [comment]
Right. Kotlin's `List` is read-only, so it can be _and is_ covariant in its type parameter: `interface List<out E>`. Since Java/JVM only support use-site variance, usages of that type are represented in bytecode as `List<? extends Whatever>`. Which is an invalid injection point type.

On the other hand, Kotlin's `MutableList` is mutable, so it must be (and is) invariant: `interface MutableList<E>`. In bytecode, usages of that type look like `List<Whatever>`. And that is a valid injection point type.

Technically, I think we could make a special case for this situation, but I'll leave that decision up to @mkouba.

**Comment by mkouba** [comment]
> Technically, I think we could make a special case for this situation, but I'll leave that decision up to @mkouba.

Like relax the spec requirements for an injetion point type in this particular case? I'm not a big fan of similar things.

**Comment by kdubb** [comment]
I'm not sure how to achieve it but a special case for this situation would be greatly appreciated; every time I use `@All` I trip over this first haha.

**Comment by Dudeplayz** [comment]
I don't know how complex it is, but would it be possible to add a Kotlin compiler plugin, which fixes such particular places?

Also, a mention in the docs somewhere that this is a problem, would be helpful to reduce and avoid this pitfall.

**Comment by geoand** [comment]
The best we can do without having to go to great lengths is provide an actionable error message

**Comment by mkouba** [comment]
> The best we can do without having to go to great lengths is provide an actionable error message

Hm, kotlin classes are annotated with a special annotation, right?

**Comment by Ladicek** [comment]
Yes, Kotlin-generated classes have the `kotlin.Metadata` annotation. ArC already uses it on one place (see `SubclassGenerator`).
